### PR TITLE
Docs: Initial generation of docs from code comments

### PR DESCRIPTION
**Pull request by haganbmj**
This is unfinished, but I wanted to at least open up a PR as a proof of concept and to see if this is worth finishing.

Reads from source files and generates the protocol document from that using a mustache template.

Proposed TODOs:
- Document the rest of the requests/events.
- ~~Child properties of response json objects (such as in the case of Stream settings).~~
- ~~Travis script to generate docs and publish them to gh-pages.~~ (edit: publish back to branch)
- Better categorization and sorting of request/event types.
- General code cleanup (which would probably involve dumping this into its own folder too).

Related to #44

### CODE REVIEW

File: docs/comments.json [hunk 1]
```diff
@@ -0,0 +1,754 @@
+{
```
Comment (haganbmj) [inline docs/comments.json]: This file is generated from the code comments.

File: docs/protocol.md [hunk 2]
```diff
@@ -0,0 +1,293 @@
+<!DOCTYPE md>
```
Comment (haganbmj) [inline docs/protocol.md]: This file is generated based on protocol.mustache and comments.json

File: WSRequestHandler.cpp [hunk 3]
```diff
@@ -1,21 +1,21 @@
...
-MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-GNU General Public License for more details.
-
-You should have received a copy of the GNU General Public License along
-with this program. If not, see <https://www.gnu.org/licenses/>
-*/
+/**
+ * obs-websocket
+ * Copyright (C) 2016-2017	Stéphane Lepin <stephane.lepin@gmail.com>
+ * Copyright (C) 2017	Brendan Hagan <https://github.com/haganbmj>
```
Comment (haganbmj) [inline WSRequestHandler.cpp]: Whoops, didn't mean to do that. Blindly copied this over from the other file. I'll revert when I have a chance.

**Comment by Palakis** [comment]
Thanks a lot for this! That last commit is me fixing the mess caused in PRs caused by my recent code style refactor.

**Comment by haganbmj** [comment]
No worries, do you have a outline of the code style you want to follow?

**Comment by Palakis** [comment]
The only two guidelines for code style are:
- Function blocks and conditionals: don't dedicate a line for a block's opening bracket (it was the case before, but I changed my mind)
- Variables: Align the `*` pointer symbol with the type, not the name

**Comment by Palakis** [comment]
> General code cleanup (which would probably involve dumping this into its own folder too).

Did you see anything shocking or out of place? :smile:

**Comment by Palakis** [comment]
@haganbmj I'd like to help you on this. Have you enabled the "Allow edits from maintainers" option?

**Comment by haganbmj** [comment]
I believe so, you were able to merge to it. I'm on mobile now, I can check later. 
I think I'll try to locate all of this into just a docs folder so it's not as much clutter.

**Comment by haganbmj** [comment]
Yeah maintainer edits are enabled.

**Comment by haganbmj** [comment]
Alright, tidied this up a bit.
Rebased on latest master, moved all the docs related stuff into the `docs` directory.
Split the js into two files and added a brief readme.

What I'd love is to get the rest of the methods/events commented and a travis script that generates docs on every commit, I'm thinking it might be best to just push changes back to master since that's where you're releasing from.

- Commit, travis generates docs after a build.
- If differences exist, push the doc changes back to master.
- Some logic to avoid wasting build time on travis commits maybe.

Mustache is a bit limiting due to how it needs the object to be structured (only being able to iterate arrays is killer), so I might see if there's any alternative or some simple way to overcome that.

**Comment by haganbmj** [comment]
Alright, so this will generate docs and push them back to master assuming you have a `GH_TOKEN` in your travis config.
I'm running into some issues with line endings though. It definitely appears to be ignoring some line endings during the build on OSX which is causing some differences in the output there.

My only other issue is that whenever this commits back to master it triggers a full rebuild in both travis and appveyor - not sure if that's really desirable since the protocol doc isn't included in a deployment.

**Comment by haganbmj** [comment]
Eh, "fixed" the line ending issue by just stripping out line numbers and such from the generated files.
Split the documentation build off into its own travis job.
Not sure why the OSX builds are having trouble getting started, probably a travis issue. It was happening all evening.

**Comment by haganbmj** [comment]
osx builds are hanging again.
Switched to handlebars. Cleaner templating. 
All that's really left is to document everything.

**Comment by haganbmj** [comment]
Everything's documented.
I'll rebase after #109 gets merged and update the docs for that, then this should be ready to go.

**Comment by Palakis** [comment]
> I'm thinking it might be best to just push changes back to master since that's where you're releasing from.

> My only other issue is that whenever this commits back to master it triggers a full rebuild in both travis and appveyor - not sure if that's really desirable since the protocol doc isn't included in a deployment.

I think it would be better to have documentation on GitHub Pages. It will avoid triggering a new build if I add exclusions on Travis and AppVeyor for the GH Pages branch. But I'll take care of this on my end, you've already done amazing work!

**Comment by haganbmj** [comment]
Yeah no worries. I'll get to it in 12 hours or so.

**Comment by haganbmj** [comment]
So the alternative is that I just have travis ignore its own commits, which I think is a cleaner solution in this case. 
I'd rather leave the Protocol in master in this case so that you can have your tags point to both the source files and the documentation? Adding an entire branch for just one file seems unnecessary.

Rebased, added the SyncOffset documentation, updated Travis to not trigger on doc commits.
I squashed commits, but I'd still recommend either Squash and Merge or Rebase and Merge instead of generating a Merge Commit.

I also purposefully didn't include the updated Protocol with SyncOffset so that it would trigger a travis build. Double check that you have `GH_TOKEN` as an environment variable in travis.

File: CI/generate-docs.sh [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    echo "-- Skipping documentation deployment because this is either a pull request or a non-master branch."
+    exit 0
+fi
+
+REMOTE_URL="$(git config remote.origin.url)"
+TARGET_REPO=${REMOTE_URL/https:\/\/github.com\//github.com/}
+GITHUB_REPO=https://${GH_TOKEN:-git}@${TARGET_REPO}
+
+git config user.name "Travis CI"
+git config user.email "$COMMIT_AUTHOR_EMAIL"
```
Comment (Palakis) [inline CI/generate-docs.sh]: Do I have to add this environment variable to Travis as well?

Comment (haganbmj) [inline CI/generate-docs.sh]: This is pulled from the commit that started the build.

**Pull request merged**